go-controller/pkg/ovnwebhook/nodeadmission.go

```diff
@@ -0,0 +1,169 @@
+		if v.action == removed {
+			return fmt.Errorf("%s cannot be removed", util.OvnNodeChassisID)
+		}
+		if v.action == changed {
+			return fmt.Errorf("%s cannot be changed once set", util.OvnNodeChassisID)
+		}
+		return nil
+	},
+	util.OvnNodeZoneName: func(v annotationChange, nodeName string) error {
+		// it is allowed for the annotation to be set to "global" or <nodeName> initially
+		if v.action == added && (v.value == types.OvnDefaultZone || v.value == nodeName) {
+			return nil
+		}
+
+		// it is allowed for the annotation to be changed to <nodeName>
+		if v.action == changed && v.value == nodeName {
+			return nil
+		}
+
+		return fmt.Errorf("%s can only be set to %s or %s, it cannot be removed", util.OvnNodeZoneName, types.OvnDefaultZone, nodeName)
+	},
+}
+
+// interconnectNodeAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users in IC environments
+var interconnectNodeAnnotationChecks = map[string]checkNodeAnnot{
+	util.OvnNodeMigratedZoneName: func(v annotationChange, nodeName string) error {
+		// it is allowed for the annotation to be set to <nodeName>
+		if (v.action == added || v.action == changed) && v.value == nodeName {
+			return nil
+		}
+
+		return fmt.Errorf("%s can only be set to %s, it cannot be removed", util.OvnNodeMigratedZoneName, nodeName)
+	},
+}
+
+// hybridOverlayNodeAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users hybrid overlay environments
+var hybridOverlayNodeAnnotationChecks = map[string]checkNodeAnnot{
+	hotypes.HybridOverlayDRMAC: nil,
+	hotypes.HybridOverlayDRIP:  nil,
+}
+
+type NodeAdmission struct {
+	annotationChecks map[string]checkNodeAnnot
+	annotationKeys   sets.Set[string]
+}
+
+func NewNodeAdmissionWebhook(enableInterconnect, enableHybridOverlay bool) *NodeAdmission {
+	checks := make(map[string]checkNodeAnnot)
+	maps.Copy(checks, commonNodeAnnotationChecks)
+	if enableInterconnect {
+		maps.Copy(checks, interconnectNodeAnnotationChecks)
+	}
+	if enableHybridOverlay {
+		maps.Copy(checks, hybridOverlayNodeAnnotationChecks)
+	}
+	return &NodeAdmission{
+		annotationChecks: checks,
+		annotationKeys:   sets.New[string](maps.Keys(checks)...),
+	}
+}
+
+var _ admission.CustomValidator = &NodeAdmission{}
+
+func (p NodeAdmission) ValidateCreate(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateDelete(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore deletion, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldNode := oldObj.(*corev1.Node)
+	newNode := newObj.(*corev1.Node)
+
+	changes := mapDiff(oldNode.Annotations, newNode.Annotations)
```

Comment:
changes and changedKeys are not used until line 115. Shall we postpone doing mapDiff until then,
by moving these 2 lines below?